gateway/src/endpoints/batch_inference.rs

```diff
@@ -0,0 +1,886 @@
+                function,
+                &inference_configs,
+                &inference_clients,
+                variant_inference_params,
+            )
+            .await;
+
+        let result = match result {
+            Ok(result) => result,
+            Err(e) => {
+                tracing::warn!(
+                        "functions.{function_name}.variants.{variant_name} failed during inference: {e}",
+                        function_name = params.function_name,
+                        variant_name = variant_name,
+                    );
+                variant_errors.insert(variant_name.to_string(), e);
+                continue;
+            }
+        };
+
+        // Write to ClickHouse (don't spawn a thread for this because it's required)
+        let write_metadata = BatchInferenceDatabaseInsertMetadata {
+            function_name: params.function_name.as_str(),
+            variant_name,
+            episode_ids: &episode_ids,
+            tags: params.tags,
+        };
+
+        let (batch_id, inference_ids) = write_inference(
+            &clickhouse_connection_info,
+            params.inputs,
+            result,
+            write_metadata,
+            inference_config.clone(),
+            // Spent a while fighting the borrow checker here, gave up
+            // The issue is that inference_config holds the ToolConfigs and ModelInferenceRequest has lifetimes that conflict with the inference_config
+        )
+        .await?;
+
+        return Ok(Json(PrepareBatchInferenceOutput {
+            batch_id,
+            inference_ids,
+            episode_ids,
+        })
+        .into_response());
+    }
+
+    // Eventually, if we get here, it means we tried every variant and none of them worked
+    Err(ErrorDetails::AllVariantsFailed {
+        errors: variant_errors,
+    }
+    .into())
+}
+
+#[derive(Debug, Serialize)]
+struct PrepareBatchInferenceOutput {
+    batch_id: Uuid,
+    inference_ids: Vec<Uuid>,
+    episode_ids: Vec<Uuid>,
+}
+
+#[derive(Debug)]
+struct BatchInferenceDatabaseInsertMetadata<'a> {
+    pub function_name: &'a str,
+    pub variant_name: &'a str,
+    pub episode_ids: &'a Vec<Uuid>,
+    pub tags: Option<Vec<Option<HashMap<String, String>>>>,
+    // pub tool_configs: &'a Vec<Option<ToolCallConfig>>,
+}
+
+#[derive(Debug, Serialize)]
+struct BatchModelInferenceInsert<'a> {
+    pub inference_id: String,
+    pub batch_id: &'a str,
+    pub function_name: &'a str,
+    pub variant_name: &'a str,
+    pub episode_id: String,
+    pub input: String,
+    pub input_messages: String,
+    pub system: Option<&'a str>,
+    pub tool_params: Option<ToolCallConfigDatabaseInsert>,
+    pub inference_params: &'a InferenceParams,
+    pub output_schema: Option<String>,
+    pub model_name: &'a str,
+    pub model_provider_name: &'a str,
+    pub tags: HashMap<String, String>,
+}
+
+#[derive(Debug, Serialize)]
+struct BatchRequestInsert<'a> {
+    batch_id: &'a str,
+    id: String,
```

Comment:
Why do we need this `id`? Can't we just use `batch_id`?

Reply:
this is gonna get appended to with more writes as we poll so yes